include/ginkgo/core/base/range_accessors.hpp

```diff
@@ -186,6 +216,647 @@ class row_major {
+                                               storage_stride_type stride)
+        : size_{size}, storage_{storage}, stride_(stride)
+    {}
+
+    /**
+     * Creates the accessor for an already allocated storage space with a
+     * stride. The first stride is used for computing the index for the first
+     * index, the second stride for the second index, and so on.
+     *
+     * @param storage  pointer to the block of memory containing the storage
+     * @param size  multidimensional size of the memory
+     * @param strides  strides used for memory accesses
+     */
+    template <typename... Strides>
+    GKO_ATTRIBUTES constexpr reduced_row_major(gko::dim<dimensionality> size,
+                                               storage_type *storage,
+                                               Strides &&... strides)
+        : reduced_row_major{
+              size, storage,
+              storage_stride_type{{std::forward<Strides>(strides)...}}}
+    {
+        static_assert(sizeof...(Strides) + 1 == dimensionality,
+                      "Number of provided Strides must be dimensionality - 1!");
+    }
+
+    /**
+     * Creates the accessor for an already allocated storage space.
+     * It is assumed that all accesses are without padding.
+     *
+     * @param storage  pointer to the block of memory containing the storage
+     * @param size  multidimensional size of the memory
+     */
+    GKO_ATTRIBUTES constexpr reduced_row_major(dim<dimensionality> size,
+                                               storage_type *storage)
+        : reduced_row_major{size, storage,
+                            helper::compute_stride_array<size_type>(size)}
+    {}
+
+    /**
+     * Creates an empty accessor (pointing nowhere with an empty size)
+     */
+    GKO_ATTRIBUTES constexpr reduced_row_major()
+        : reduced_row_major{{0, 0, 0}, nullptr}
+    {}
+
+public:
+    /**
+     * Creates a reduced_row_major range which contains a read-only version of
+     * the current accessor.
+     *
+     * @returns  a reduced_row_major major range which is read-only.
+     */
+    GKO_ATTRIBUTES GKO_INLINE constexpr range<const_accessor> to_const() const
+    {
+        return range<const_accessor>{size_, storage_, stride_};
+    }
+
+    /**
+     * Returns the length in dimension `dimension`.
+     *
+     * @param dimension  a dimension index
+     *
+     * @returns  length in dimension `dimension`
+     */
+    GKO_ATTRIBUTES GKO_INLINE constexpr size_type length(
+        size_type dimension) const
+    {
+        return dimension < dimensionality ? size_[dimension] : 1;
+    }
+
+    /**
+     * Copies data from another accessor
+     *
+     * @warning Do not use this function since it is not optimized for a
+     *          specific executor. It will always be performed sequentially.
+     *          Please write an optimized version (adjusted to the architecture)
+     *          by iterating through the values yourself.
```

Comment:
It doesn't have to be part of the PR, but it would be nice to have an example on that topic at some point, I think.